WiP: Redesign delete/remove dialog

skobyda opened the pull request:
As pointed out at https://github.com/cockpit-project/cockpit-machines/pull/640 , there are many delete/removal dialogs in c-machines which do not follow PF guidelines. Here is my initial try to adhert to the guidelines. It will likely need some rewording.

- I mostly moved identification of what we are removing from modal header, where it sometimes may not fit, into the modal body.
- Fixed some Upper Case in Network deletion modal.
- Implemented DescriptionList for more complex object which cannot be identified by just 1 value

### Before
![Screenshot from 2022-04-04 13-58-28](https://user-images.githubusercontent.com/42733240/161732899-c9b546ed-2593-45f2-8b5f-249255632d14.png)

### After
![Screenshot from 2022-04-05 11-43-31](https://user-images.githubusercontent.com/42733240/161732930-4fa33f9b-2241-4f59-90dc-f9b146a2192f.png)

### Before
![Screenshot from 2022-04-04 15-44-57](https://user-images.githubusercontent.com/42733240/161733040-fbefe5fb-b31e-46a3-892a-aee54544d187.png)

### After
![Screenshot from 2022-04-04 15-38-53](https://user-images.githubusercontent.com/42733240/161732996-977f4bee-409a-4f78-8a63-9a158255e92c.png)

### Before
![Screenshot from 2022-04-05 12-03-16](https://user-images.githubusercontent.com/42733240/161733084-8c3d9379-fbd6-4c3a-9b34-c0744f7e5c65.png)

### After
![Screenshot from 2022-04-05 12-03-47](https://user-images.githubusercontent.com/42733240/161733121-356529e4-2d29-4e97-a659-7e0d270b4c2e.png)

### Before
![Screenshot from 2022-04-05 12-03-54](https://user-images.githubusercontent.com/42733240/161733169-4f2a0197-f747-4aa2-a113-600c23d6c56c.png)

### After
![Screenshot from 2022-04-05 12-03-59](https://user-images.githubusercontent.com/42733240/161733205-c7511fdb-6488-4a16-a5ec-587874ca3f5c.png)

### Before
![Screenshot from 2022-04-05 12-09-09](https://user-images.githubusercontent.com/42733240/161733245-c73cfa09-09c4-490d-9d13-1f30d7e2b066.png)

### After
![Screenshot from 2022-04-05 12-12-09](https://user-images.githubusercontent.com/42733240/161733269-4cda7f71-4c58-4fa3-bd6f-6f9159b0cce2.png)

skobyda commented:
I'm also thinking how to redesign Storage Pool deletion dialog. As it (depending on whetever is active) contains additional message or checkbox:
![Screenshot from 2022-04-05 12-15-25](https://user-images.githubusercontent.com/42733240/161734205-51f4d838-6109-4aec-bc0d-3c35655f0cdd.png)
![Screenshot from 2022-04-05 12-13-50](https://user-images.githubusercontent.com/42733240/161734209-2907a848-f15a-4f67-a90d-cd6dc91617a3.png)

garrett commented:
> I'm also thinking how to redesign Storage Pool deletion dialog. As it (depending on whetever is active) contains additional message or checkbox:
> ![Screenshot from 2022-04-05 12-27-09](https://user-images.githubusercontent.com/42733240/161734495-ff622d6c-13c8-4e11-b2c8-6b77aceff72d.png)
> ![Screenshot from 2022-04-05 12-26-52](https://user-images.githubusercontent.com/42733240/161734497-71b71bf9-9c37-48cc-8be0-6ef293030db7.png)

I don't understand. Is this what it currently looks like?

There shouldn't be a label here, except for the mandatory one for the checkbox. Which volumes would optionally be removed? It should show those.

For example, when I expand my default storage pool, I see this:

![image](https://user-images.githubusercontent.com/10246/162002233-f7a3cc04-9fb7-48cb-b405-5ddc66f5f148.png)

Can you even delete the storage pool without the volumes? :thinking: 

...What would happen to the files? And the VMs?

...And then what happens to the VMs when you delete the storage pool and the volumes?

KKoukiou reviewed (commented):
I won't manage to review this PR. Just from the screen shots, I noticed that the modal titles don't follow the PF guidelines. In specfiic, Confirmation modals, need to have titles with a question.
For example;

https://www.patternfly.org/v4/components/modal/design-guidelines/#confirmation-dialogs

skobyda commented:
> I don't understand. Is this what it currently looks like?

Yes, that's the current look.

> There shouldn't be a label here, except for the mandatory one for the checkbox. Which volumes would optionally be removed? It should show those.
> 
> For example, when I expand my default storage pool, I see this:
> 
> ![image](https://user-images.githubusercontent.com/10246/162002233-f7a3cc04-9fb7-48cb-b405-5ddc66f5f148.png)
> 
> Can you even delete the storage pool without the volumes? thinking
>
> ...What would happen to the files? And the VMs?

You can delete the pool without volumes. In that case, the files (e.g. qcow2 file or raw file) are left untouched. For example, if you have a file directory storage pool, then deleting a storage pool just deletes the storage pool "abstraction" over that directory, but the directory with files is still left on the system. This can be useful if you want to move some of those files/disk elsewhere after storage pool deletion.

> ...And then what happens to the VMs when you delete the storage pool and the volumes?

the storage pool **OR** the volumes. Those are 2 different situations:

If some VM is using disk with reference as [storagePoolName]:[volumeName], then you cannot delete the storage pool:
_(Button "Delete" should be probably greyed out. This needs aria-disabled label!)_
![Screenshot from 2022-04-07 14-13-00](https://user-images.githubusercontent.com/42733240/162197152-94fccbad-a73b-483b-aa56-06ac6771958c.png)

If you reference the storage volume directly thru path or some other ID without referencing storage pool, then you can delete the storage pool:
_(Maybe those info text shoould be a tooltip of disabled Button)_
![Screenshot from 2022-04-07 14-20-55](https://user-images.githubusercontent.com/42733240/162197358-f9775064-6364-4f7a-946f-2ee46c7ed05d.png)

skobyda commented:
> I won't manage to review this PR. Just from the screen shots, I noticed that the modal titles don't follow the PF guidelines. In specfiic, Confirmation modals, need to have titles with a question. For example;
> 
> https://www.patternfly.org/v4/components/modal/design-guidelines/#confirmation-dialogs

Then maybe just rewording those titles should be enough:
"Remove network" ->"Remove network?"
"Remove disk from VM" -> "Remove disk from VM?"

garrett commented:
I've mentioned our headlines not conforming to PF titles before (regarding the "?"), but we never did it as it was a one-off and inconsistent with the rest of Cockpit.

As this is one of those PRs with sweeping changes, it makes sense to switch the titles to questions wherever it makes sense. :+1: 

---

If we're making these changes, we might want to also consider adding an icon for destruction too:

https://www.patternfly.org/v4/components/modal/design-guidelines/#confirm-a-destructive-action
![image](https://user-images.githubusercontent.com/10246/162225900-73681f6c-893d-441b-9828-262fb2cde033.png)

> Use a [primary button](https://www.patternfly.org/v4/components/button/design-guidelines#primary-button) to confirm a destructive action. If the action carries serious consequences, then use a [danger button](https://www.patternfly.org/v4/components/button/design-guidelines#danger-button) instead.
> 
> If an action is **difficult or impossible to undo, add a warning icon** beside the headline.

They're optional... But it seems to strongly suggest them above under certain destructive circumstances.

https://www.patternfly.org/v4/components/modal/design-guidelines/#icon-use-in-modal-dialogs:

> <p class="ws-p">Icons are optional in modal dialogs. Use or omit them as your use case requires.</p>
> 
>  Use case(s) | Usage
>  -- | --
>   Warning: Cautions or warns the user of a permanent action, or that information will be deleted upon action completion | Add to confirmation dialogs or passive dialogs to indicate a higher level of urgency and importance.
>    Critical Warning: Indicates that an error has occured | Add to error dialogs.
>    Acknowledgement: Informs the user of an action or result | Add to confirmation or passive dialogs to indicate a lower level of urgency.
> 

(But this could easily go in another PR.)

garrett commented:
> If some VM is using disk without storage pool reference, but instead is referencing the storage volume directly thru file path or some other ID without referencing storage pool, then you can delete the storage pool, but not its volumes:
(Maybe that info text shoould be a tooltip of disabled Button?)
>
> ![image](https://user-images.githubusercontent.com/10246/162228541-93ab0ceb-3347-4bd0-82ec-4417fd40cdbd.png)

Should that dialog have some sort of solution within the dialog itself?

That is, list the VMs it's in use by and perhaps have a way to detach them from the individual VMs from within the dialog?

(This probably should be addressed in an separate PR; keep this PR mainly related to reworking the modals as-is without really adding features — except for showing more information.)

garrett commented:
What happens with errors? Are they inline alerts? Are those at the bottom or top? (If inline alerts, they should be at the top, like what https://github.com/cockpit-project/cockpit/pull/17221 is now doing.)

I know that Machines does do a lot of error handling as alerts in the page, after a modal though, so perhaps it doesn't even need to be addressed. :grin:

skobyda commented:
> it makes sense to switch the titles to questions wherever it makes sense. +1

Done

> If we're making these changes, we might want to also consider adding an icon for destruction too

Done

![Screenshot from 2022-04-20 21-45-25](https://user-images.githubusercontent.com/42733240/164320942-1543447d-f323-4709-aa1c-66e3579b78af.png)

> > If some VM is using disk without storage pool reference, but instead is referencing the storage volume directly thru file path or some other ID without referencing storage pool, then you can delete the storage pool, but not its volumes:
> > (Maybe that info text shoould be a tooltip of disabled Button?)
> > ![image](https://user-images.githubusercontent.com/10246/162228541-93ab0ceb-3347-4bd0-82ec-4417fd40cdbd.png)
> 
> Should that dialog have some sort of solution within the dialog itself?
> 
> That is, list the VMs it's in use by and perhaps have a way to detach them from the individual VMs from within the dialog?
> 
> (This probably should be addressed in an separate PR; keep this PR mainly related to reworking the modals as-is without really adding features — except for showing more information.)

Yeah

> What happens with errors? Are they inline alerts? Are those at the bottom or top? (If inline alerts, they should be at the top, like what [cockpit-project/cockpit#17221](https://github.com/cockpit-project/cockpit/pull/17221) is now doing.)
> 
> I know that Machines does do a lot of error handling as alerts in the page, after a modal though, so perhaps it doesn't even need to be addressed. grin

They were inline errors in all the cases, except for disk removal, where there were alerts in page (but that was accidental, I tracked it down to my [old PR](https://github.com/cockpit-project/cockpit/pull/12661) where I forgot to replace page alert there with inline modal alert.

So now I made it consistent, all the delete/remove dialogs now use inline alerts, and fixed them to show alert at the top of the modal:

![Screenshot from 2022-04-20 23-00-24](https://user-images.githubusercontent.com/42733240/164322038-b4dffbdf-04f6-4ed2-97a7-0ec669976d3f.png)

garrett reviewed (changes requested):
> > it makes sense to switch the titles to questions wherever it makes sense. +1
> 
> Done
> 
> > If we're making these changes, we might want to also consider adding an icon for destruction too
> 
> Done
> 

Nice! Thanks!

---

Anything we can do about this error message?

![image](https://user-images.githubusercontent.com/10246/164473604-0b6c3058-e203-4aba-98da-ad105cf51448.png)

1. We should never use the title "Failure". It should be specific to the error type.
2. What happened to the space between the alert and the rest of the modal content? (There is none here; there should be some.)
3. A CD is not a "disk". A CD is a "disc". (It's a "disc" with a C if it's a circle you can see.)
4. Should "target" really be "device"?
   For example. GNOME Disks labels the it as a "device":
   ![image](https://user-images.githubusercontent.com/10246/164474645-d1aad97d-dfb1-4a6d-b6a1-a913765e23a3.png)

---

> ![image](https://user-images.githubusercontent.com/10246/164475076-45d53f77-55d2-4f20-96d2-acd0d4f0acf3.png)
> ...
> Yeah. This functionality would be way more complex, so I would leave it outside of this PR.

Did you add the checkbox's label here? It shouldn't have that duplicate bold label to the left... It's a single opt-in checkbox. It also shouldn't have "the": "Delete volumes inside this pool". Additionally, the whole "c..." thing is odd. But I understand that part will be handled in a follow-up.

skobyda commented:
> Anything we can do about this error message?
> 
> ![image](https://user-images.githubusercontent.com/10246/164473604-0b6c3058-e203-4aba-98da-ad105cf51448.png)
> 
>     1. We should never use the title "Failure". It should be specific to the error type.
>
>     2. What happened to the space between the alert and the rest of the modal content? (There is none here; there should be some.)

There was no by default in ModalError component. I added patternfly's padding-top medium: `pf-u-pt-md` 

>     3. A CD is not a "disk". A CD is a "disc". (It's a "disc" with a C if it's a circle you can see.)

Fixed it in modal title and error message title. However it's still present in error message, since that error message comes from libvirt itself.

![Screenshot from 2022-04-26 01-21-51](https://user-images.githubusercontent.com/42733240/165191379-b2921782-809b-4712-bcaa-45b011302f92.png)


>     4. Should "target" really be "device"?
>        For example. GNOME Disks labels the it as a "device":
>        ![image](https://user-images.githubusercontent.com/10246/164474645-d1aad97d-dfb1-4a6d-b6a1-a913765e23a3.png)

Term "target" is quite commonly used across libvirt, and there is reason for it, as "device" and "target" may refer to different things.

- "Device" is the identifier under which the disk is known on the host.
- "Target" specifies identifier which will be exposed to the VM. (i.e. how it's "targeted" to the VM)

Those 2 will be usually different. 
For example: you have "sda" device on your host, you attach it to the VM, but device will be exposed to the VM under identifier "vda"
I believe if we just call it "device" , we eventually end up confusing user, and they might think they have attached some "vdb" device from host to the VM.
 
> Did you add the checkbox's label here? It shouldn't have that duplicate bold label to the left... It's a single opt-in checkbox. It also shouldn't have "the": "Delete volumes inside this pool". Additionally, the whole "c..." thing is odd. But I understand that part will be handled in a follow-up.

No, the label was already there.

garrett commented:
> What do you mean? There was never a space in our AlertNotification component.

I meant the space in that screenshot. There should be space; alert widgets should never be slammed into something else. (And most other widgets shouldn't either.)

Disc: Nice! Thanks for fixing that (in Cockpit, where possible).

Target: Ok. That's fine.

skobyda commented:
Current state of this PR:
![Screenshot from 2022-04-27 12-33-39](https://user-images.githubusercontent.com/42733240/165499592-267ce721-3d34-4735-a10f-5a6a6cad21a6.png)
![Screenshot from 2022-04-27 12-32-53](https://user-images.githubusercontent.com/42733240/165499603-edfc63a1-3a29-4187-a665-b4c4a7c701c3.png)
![Screenshot from 2022-04-27 12-32-40](https://user-images.githubusercontent.com/42733240/165499610-1cf2c67c-7668-475b-9dc3-cb702aac45fe.png)
![Screenshot from 2022-04-27 12-32-04](https://user-images.githubusercontent.com/42733240/165499614-3a566abf-90dc-4165-a445-eac4fdb320f6.png)
![Screenshot from 2022-04-27 12-31-43](https://user-images.githubusercontent.com/42733240/165499619-e23ee532-0f82-4c1a-849d-8e1a0c607494.png)
![Screenshot from 2022-04-27 12-31-28](https://user-images.githubusercontent.com/42733240/165499627-c5574956-95ff-4004-9d94-7a10a113cdf6.png)
![Screenshot from 2022-04-27 12-30-59](https://user-images.githubusercontent.com/42733240/165499632-8f4b7c01-2c65-4338-9dd0-95895adfd6f7.png)
![Screenshot from 2022-04-27 12-30-47](https://user-images.githubusercontent.com/42733240/165499635-c2741d15-89b3-4c96-a80a-f26b6b6b7cd1.png)
![Screenshot from 2022-04-27 12-30-32](https://user-images.githubusercontent.com/42733240/165499643-4c6c9ca0-f4e2-410a-a117-4d4a26c24a19.png)

skobyda commented:
The last issues, which can be seen on screenshots above, is the deletion of storage pool, which will probably need more comprehensive redesign:
![Screenshot from 2022-04-27 12-33-39](https://user-images.githubusercontent.com/42733240/165499836-36490212-e153-42c2-a98d-a1b18d1d019a.png)

And deleteting VMs, which it seems doesn't mention the VM name anywhere in the dialog. Maybe it should. Is it okay to be mentioned in the title, or should the modal message be reworded to include the VM name?
![Screenshot from 2022-04-27 12-30-32](https://user-images.githubusercontent.com/42733240/165499851-8d093313-83fe-416c-b530-7d4b838b2896.png)

garrett reviewed (changes requested):
Most of these changes are excellent! Thanks!

What's up with the `:` here in this dialog?

![image](https://user-images.githubusercontent.com/10246/165549142-62aba268-e628-41f7-b814-915bbda06a10.png)


This still shouldn't have a label at the side, and it still doesn't mention what will be deleted. This is not great. :cry: 

![image](https://user-images.githubusercontent.com/10246/165548799-97db3f38-fba9-4074-9bb7-8bcffd8ea23c.png)


What's up with the spacing? It should look more like this:

![image](https://user-images.githubusercontent.com/10246/165547360-4e0c8f11-b52a-409a-a168-79c25b73a478.png)

marusak commented on src/components/common/deleteResource.jsx:
```diff
@@ -63,10 +64,12 @@ export class DeleteResourceModal extends React.Component {
             <Modal position="top" variant="small" isOpen onClose={onClose}
                    id="delete-resource-modal"
-                   title={title}
+                   title={<>
```
Modal has `titleIconVariant` property where you can specify either `warning` or specific icon, see https://www.patternfly.org/v4/components/modal

Cannot we use that instead?

skobyda replied:
Good catch, thanks!

skobyda commented:
@garrett So here is my new iteration:
### Adjusted spacing in VM deletion dialog:
![Screenshot from 2022-05-03 14-40-11](https://user-images.githubusercontent.com/42733240/166455509-2d0526d2-8f79-48ce-ad01-6b408a91b907.png)

### Storage pool with volumes will mention which volumes will be deleted:
![Screenshot from 2022-05-03 14-43-00](https://user-images.githubusercontent.com/42733240/166455618-b1f9ce62-65f6-46c7-af44-7df9913a91dc.png)

### Inactive storage pool:
![Screenshot from 2022-05-03 14-43-53](https://user-images.githubusercontent.com/42733240/166455697-c71c403f-c830-4529-97a8-8de8ddb669ce.png)

### Storage pool without any volumes:
This however looks a bit strange, to have just an empty modal. Or not? Not sure about this one.
![Screenshot from 2022-05-03 14-44-00](https://user-images.githubusercontent.com/42733240/166455757-c080924e-62d7-434f-bf0d-d52f26e975e9.png)

skobyda commented:
> ![image](https://user-images.githubusercontent.com/10246/165548799-97db3f38-fba9-4074-9bb7-8bcffd8ea23c.png)

User would want to delete pool without deleting volumes if they are migrating them to another volumes. E.g:
1. Have a directory pool at /var/lib/libvirt/myPool1 with volumes as file in it.
2. You want to move those volumes to another pool, let's say /var/lib/libvirt/myPool2
3. To do that, you would delete myPool1, and then just move it's leftover volumes to myPool2 directory

martinpitt commented on src/components/vm/disks/vmDiskColumns.jsx:
```diff
@@ -46,15 +65,9 @@ export const DiskSourceCell = ({ diskSource, idPrefix }) => {
     };
 
     const chunks = [];
-    addOptional(chunks, diskSource.file, "file", _("File"));
-    addOptional(chunks, diskSource.dev, "device", _("Device"));
-    addOptional(chunks, diskSource.protocol, "protocol", _("Protocol"));
-    addOptional(chunks, diskSource.pool, "pool", _("Pool"));
-    addOptional(chunks, diskSource.volume, "volume", _("Volume"));
-    addOptional(chunks, diskSource.host.name, "host", _("Host"));
-    addOptional(chunks, diskSource.host.port, "port", _("Port"));
+    DISK_SOURCE_LIST.forEach(entry => addOptional(chunks, getDiskSourceValue(diskSource, entry.name), entry.name, entry.label));
 
-    return <DescriptionList isHorizontal>{chunks}</DescriptionList>;
+    return <DescriptionList isHorizontal>{chunks.map(chunk => chunk)}</DescriptionList>;
```
This looks like a no-op.

martinpitt commented on src/components/vm/disks/vmDisksCard.jsx:
```diff
@@ -267,12 +267,24 @@ export class VmDisksCard extends React.Component {
-                title: cockpit.format(_("Remove disk $0"), disk.target),
-                errorMessage: cockpit.format(_("Disk $0 could not be removed"), disk.target),
+                errorMessage: disk.device === "cdrom"
+                    ? cockpit.format(_("Disc $0 could not be removed"), disk.target)
+                    : cockpit.format(_("Disk $0 could not be removed"), disk.target),
```
[My dictionary](https://dict.leo.org/englisch-deutsch/disc) says that disc vs. disk are mostly just American vs. British English. Duplicating these strings seems a bit dubious and unnecessary, especially as they are all in the same "Disks" table on the VM details anyway.

skobyda replied:
@garrett mind giving input?

skobyda replied:
Yeah, but it seems there are specific cases where it's used independed of variation of english. And that is optical discs:
1. `Optical disc image` or in short: `ISO`
It sems that even wikipedia calls it "disk" when it's a disk device in general, but "disc" when talking about ISO/CD: An optical **disc** image (or ISO image) is a **disk** image. [wiki](https://en.wikipedia.org/wiki/Optical_disc_image)
3. CD: `Compact disc read-only memory`
4. DVD: `Digital Versatile Disc` or `Digital Video Disc`

And when we are talking about ISO images in VMs, we are talking about emulations of Optical Disc image or CD.

martinpitt replied:
Right, I'm aware of that technical distinction -- but it doesn't make much difference here in the dialog IMHO. The VM details page has them all under the "Disks" title.

skobyda replied:
Alright, reverted

martinpitt commented on src/components/vm/hostdevs/hostDevCard.jsx:
```diff
@@ -231,12 +231,32 @@ export const VmHostDevCard = ({ vm, nodeDevices, config }) => {
+                                              title: _("Remove host device from VM?"),
                                               errorMessage: ("Host device could not be removed"),
+                                              actionDescription: cockpit.format(_("Host device will be removed from $0:"), vm.name),
+                                              objectDescription,
```
Please adjust the test to make sure the dialog contains the expected device description.

skobyda replied:
Done

martinpitt commented on test/machineslib.py:
```diff
@@ -212,7 +212,7 @@ def deleteIface(self, iface):
 
         b.click(f"#delete-vm-subVmTest1-iface-{iface}")
-        b.wait_in_text(".pf-c-modal-box .pf-c-modal-box__header .pf-c-modal-box__title", "Delete network interface")
+        b.wait_in_text(".pf-c-modal-box .pf-c-modal-box__header .pf-c-modal-box__title", "Delete network interface?")
```
Please add a check for the description details in the dialog.

skobyda replied:
Virtual network static DHCP entries doesn't have any object description in the dialog, it only a sentence "network {networkMac} will be permanently deleted.". So the only thing which could go wrong is to have an incorrect network mac. But we test in other places that we have correct MAC for a network entry, so is it worth testing?

skobyda replied:
Actually, on second thought, we can never now what can go wrong, so let's lest, it. Updated the PR

martinpitt reviewed (changes requested):
Thanks! I have some minor comments and some language bikeshedding. The pixel tests need to be updated (the diff looks good). @garrett , can you please re-review the latest screenshots?

garrett reviewed (changes requested):
Overall great! 

I've made a few adjustments in the mockups; it'd be nice to get them in this PR, but I understand if some are hard to do or would otherwise be follow-ups:

1. We should probably make the storage pool "remove", not delete. Deletion is for deleting stuff on a disk, whereas removing is just removing a definition of something. So we're removing storage pools and deleting volumes.
2. If we could prioritze the name earlier in the title, it might be nice. (WDYT? Adjusted in the screenshots. Ideally, it'd be bold too. But however this is done, it should be translation-friendly.)
3. Volumes being deleted should probably be a bulleted list, not an inline comma separated list, so they have more gravitas (serious importance; visual weightiness here in the UI) and are easier to scan.
4. I've changed the text for the volumes being deleted, to state that _all_ of them in the list will be deleted.
5. When a storage pool is being removed and volumes are being deleted (where the option is checked), I have the button switch to remove & delete.
6. I added some empty pool text. It's there just to give a "peace of mind" (reassurance) that volumes aren't being deleted.
- I might be hungry. ("LunchTime" came to mind for some reason...)


![vm-pool-remove-unchecked excalidraw](https://user-images.githubusercontent.com/10246/166650501-6cbe44f5-6166-4576-95c3-56f46b2ee161.png)

![vm-pool-remove excalidraw](https://user-images.githubusercontent.com/10246/166650492-73ef26f7-7e6e-4014-a40c-830066cd2cd6.png)

![vm-pool-remove-empty excalidraw(1)](https://user-images.githubusercontent.com/10246/166651800-527fca82-c6d0-466d-87ba-d844985d9e12.png)

garrett commented:
After posting the above comment: I thought about it and storage pools are _also_ on disk, so perhaps just ignore my suggestion to change remove to delete. That'll make things easier for this PR to land. The other suggestions would still apply.

Here are modified mockups for that:

![vm-pool-delete excalidraw](https://user-images.githubusercontent.com/10246/166652875-803359fd-df05-4c9a-9e71-211adb417711.png)

![vm-pool-delete-empty excalidraw](https://user-images.githubusercontent.com/10246/166652881-709c8f1b-66d3-4779-a2df-f0c1a605c8d5.png)

skobyda commented:
Updated design:
![Screenshot from 2022-05-09 12-53-06](https://user-images.githubusercontent.com/42733240/167396715-570b1751-fa7b-45ea-b510-76dda771dd5a.png)
![Screenshot from 2022-05-09 12-52-54](https://user-images.githubusercontent.com/42733240/167396718-3dad3b9c-f6c0-46de-9211-9f1496544e55.png)

The only thing I could not achieve was to make Storage pool name in the modal title bold. The title is formatted thru cockpit.format() and it does not accept string where storage pool name is a html <strong> element. @martinpitt any idea if it's possible to highlight a word in a translated+formatted string?

martinpitt commented:
@skobyda : HTML markup inside translated strings is best avoided. There are some tricks to make it work, but it's really brittle through translations, and unsafe too.

martinpitt commented on src/components/storagePools/storagePoolDelete.jsx:
```diff
@@ -160,26 +162,34 @@ export class StoragePoolDelete extends React.Component {
-                { storagePool.active && showWarning() }
+                { storagePool.active
+                    ? (volumes.length > 0
+                        ? <>
```
No need for this extra `<>` wrapper.

martinpitt commented on test/check-machines-networks:
```diff
@@ -868,6 +868,9 @@ class TestMachinesNetworks(VirtualMachinesCase):
 
             def cleanup(self):
                 b.click(f"#delete-network-{self.network_name}-{self.connection_name}-ipv4-dhcp-host-{self.entry_id}-button")
+                b.wait_in_text(".pf-c-modal-box__body .pf-c-description-list", self.ip)
+                b.wait_in_text("#delete-resource-modal-IP", self.ip)
+                b.wait_in_text("#delete-resource-modal-MAC", self.mac)
```
This would better be squashed into the "Redesign vNIC removal dialog" commit.

skobyda replied:
This is a completely different dialog. Removing vNIC and DHCP static entries from virtual networks are different functionalities with different dialogs.

martinpitt reviewed (approved):
Thank you! Code-wise this now looks good enough. The two nitpicks are only when you need to force-push anyway for some other reason.

@garrett , are you happy with the design now?

martinpitt commented:
@skobyda: Can you please write a release note? Personally I find the effect on the "Remove host device" dialog the most obvious, so maybe take that new picture, clip it so that it has an uniform 2px border, and use it for the relnote?

garrett reviewed (changes requested):
Apologies for not catching these earlier. They're all minor things that should be able to be fixed quickly, I think?

---

![Screenshot 2022-05-11 at 10-45-19 Virtual Machines - garrett@Bolt](https://user-images.githubusercontent.com/10246/167809469-3f0758f1-b2c2-499c-b1ee-dad2548c5bbe.png)

Which VM? The modal doesn't say. (I guess this could be a follow-up.)

---

![Screenshot 2022-05-11 at 10-46-44 Virtual Machines - garrett@Bolt](https://user-images.githubusercontent.com/10246/167809474-5113ab34-875e-47e2-bf32-870600712e1c.png)

This should have a `:`

---

![Screenshot 2022-05-11 at 10-46-55 Virtual Machines - garrett@Bolt](https://user-images.githubusercontent.com/10246/167809476-4432f6a5-bf80-4f71-a5c7-dc2847ddf4c4.png)

This should not be delete. It should be remove.

![Screenshot 2022-05-11 at 10-47-14 Virtual Machines - garrett@Bolt](https://user-images.githubusercontent.com/10246/167809480-4f4bcecd-1cb9-41da-b3d1-96450a0eae5a.png)

Same for the list of networks. It should be remove, not delete.

---

![Screenshot 2022-05-11 at 10-47-58 Virtual Machines - garrett@Bolt](https://user-images.githubusercontent.com/10246/167809483-786f1307-d695-48ab-b7e9-1229fc48c0c7.png)

Missing "of":

All **of** its...

(Perhaps this should be a follow-up after the release, as it would affect translations, right?)

---

![Screenshot 2022-05-11 at 10-48-28 Virtual Machines - garrett@Bolt](https://user-images.githubusercontent.com/10246/167809487-92a9c9ca-02d7-4885-b06a-3dd739b7c941.png)

This doesn't say what it is inside the container. (I think this one's fine for a follow-up though.)

martinpitt commented:
> [network iface] This should not be delete. It should be remove.

Should it? I thought it literally ceases to exist when you del^Wrem^Wmake it go away?

martinpitt commented:
> Perhaps this should be a follow-up after the release, as it would affect translations, right?

This PR changes so many strings, it'd actually be better to not change them twice. I.e. better to change the string right away here.

garrett commented:
> Should it? I thought it literally ceases to exist when you del^Wrem^Wmake it go away?

Are files being deleted? It should be remove, not delete.

A network card that's mapped into the VM doesn't physically cease to exist when you remove it from the VM. (It can be mapped in directly or mapped in virtually, but it's still the network interface on a physical device.)

> Delete: (v.) Use to describe completely removing an object. Delete is a destructive action.
> Opposite: Create
> Notes: Erase and Remove are not recommended for this use case. See Remove for usage guidelines.

> Remove: (v.) Use to describe removing an item from a list, group, view, or other container element without completely deleting it. Also see Add and Delete.
> Opposite: Add
> Notes: If what you’re removing is not readily apparent from the context, consider adding a noun (like “Remove file”).

Also the verb pairs are create/delete, add/remove... they shouldn't really be mixed in most circumstances (as in: create/remove, add/delete).

---

> This PR changes so many strings, it'd actually be better to not change them twice. I.e. better to change the string right away here.

Yeah, that's why I was saying follow-up for many. We could just change all string changes to follow-up.

Delete to Remove technically changes a string, but we already have that translated. So: Change delete to remove for this PR, make a follow-up to land after release for the rest.

garrett commented:
Also, we need to add some text at the top for the release notes and use just one screenshot.

I _think_ I made the script work where we can just add a horizontal rule with three dashes `---` to split the content and have the release notes at top and keep the rest that's already there after it.

skobyda commented:
> This doesn't say what it is inside the container. (I think this one's fine for a follow-up though.)

What do you mean? How can we tell the user what is inside the mount?

garrett commented:
Well, if we want this to land in the upcoming release, then we can't change strings. (Changing delete to remove isn't a string change; it's done elsewhere.)

Otherwise, we can defer this to next release in two weeks.

skobyda commented:
This PR doesn't have to be int this release, so let's change the string now and leave it for the next release

skobyda commented:
Added my draft for release notes at the [first comment](https://github.com/cockpit-project/cockpit-machines/pull/652#issue-1192935992)

martinpitt commented:
Thanks @skobyda : I tweaked the language a little bit.

skobyda commented:
Ahh, more test failures

martinpitt commented:
[pixel test failure](https://logs.cockpit-project.org/logs/pull-652-20220513-131905-cc3b0593-fedora-35/log.html), it needs updating for Delete → Remove.

martinpitt commented on src/components/storagePools/storagePoolDelete.jsx:
```diff
@@ -160,26 +162,32 @@ export class StoragePoolDelete extends React.Component {
+                        ? <Checkbox id='storage-pool-delete-volumes'
+                                    isChecked={this.state.deleteVolumes}
+                                    label={<>
+                                        { _("Also delete all volumes inside this pool: ")}
```
Trailing space. This is important for i18n'ed strings, as it's basically guaranteed that this won't survive translations. If you need the space for layouting, add it *after* the gettext call.

skobyda replied:
Ah this was probably typo, as list is in a new line, there is no need for a space

martinpitt reviewed (changes requested):
One tiny issue, and the pixel test needs updating, otherwise LGTM. Thanks! Let's try and land this today, as it's very prone to bitrotting.

martinpitt commented:
@skobyda : Thanks! Please rebase, to let packit tests run.

skobyda commented:
Tests green. I think it's close to landing. @garrett @martinpitt mind giving your final word?

garrett reviewed (approved):
Looks good! Let's merge this! :+1:

martinpitt reviewed (approved):
Dakujem!
